### ISSUE TITLE
Really need help! Try horovodrun performs slower!

### ISSUE THREAD

**Issue by CXMANDTXW** [horovod/horovod#1403]
**Environment:**
1. Framework: (PyTorch)
2. Framework version:1.2.0
3. Horovod version:0.18.1
4. MPI version:4.0.1
5. CUDA version:10.0
6. NCCL version:No NCCL
7. Python version:3.7.4
8. OS and version:UBUNTU 18.04
9. GCC version:7.4.0

**Checklist:**
1. Did you search issues to find if somebody asked this question before?
2. If your question is about hang, did you read [this doc](https://github.com/horovod/horovod/blob/master/docs/running.rst)?
3. If your question is about docker, did you read [this doc](https://github.com/horovod/horovod/blob/master/docs/docker.rst)?
4. Did you check if you question is answered in the [troubleshooting guide](https://github.com/horovod/horovod/blob/master/docs/troubleshooting.rst)?

**Your question:**
Please ask your question here.
I used to use torch.nn.DataParallel(model).cuda() to run the code.Now I switched to horovod.
However,horovod is slower than DataParallel.I tested with 2 epochs,and result is
original:253s  horovod:295s
My machine has 4 gpus( 2 TITAN X，2 1080T）32 cpus
I can't understand the timeline file.
[timeline1.zip](https://github.com/horovod/horovod/files/3630962/timeline1.zip)

Can any guys help me to find the problem? Thank a lot.

**Comment by CXMANDTXW** [comment]
I found when I use horovod
gpu set 4:
[1,1]<stdout>:bacward_time:0.6748
[1,0]<stdout>:bacward_time:0.7129
[1,2]<stdout>:bacward_time:0.8603
[1,3]<stdout>:bacward_time:0.9149
[1,3]<stdout>:opt_time:0.0293(optimizer.step())
[1,0]<stdout>:opt_time:0.3149
[1,1]<stdout>:opt_time:0.3267
[1,2]<stdout>:opt_time:0.1189
[1,0]<stdout>:forward_time:0.0622
[1,2]<stdout>:forward_time:0.0788
[1,1]<stdout>:forward_time:0.0623
[1,3]<stdout>:forward_time:0.0751
gpu set 1:
[1,0]<stdout>:bacward_time:0.4495
[1,0]<stdout>:opt_time:0.0225
[1,0]<stdout>:forward_time:0.0451
use torch.nn.DataParallel instead of horovod:
forward_time:0.1886
bacward_time:0.1869
opt_time:0.0446

**Comment by tgaddair** [comment]
Hey @CXMANDTXW, judging from the large amount of time spent in Negotiation from the timeline, it looks like there's some kind of imbalance among the workers:

<img width="1384" alt="Screen Shot 2019-09-19 at 8 34 59 AM" src="https://user-images.githubusercontent.com/1742912/65258805-68e22700-dab8-11e9-92d3-945526c1f19a.png">

You'll also notice that there are large gaps between Negotiate phases with little to no time spent in Allreduce, which is suspicious:

<img width="1375" alt="Screen Shot 2019-09-19 at 8 39 26 AM" src="https://user-images.githubusercontent.com/1742912/65259135-08071e80-dab9-11e9-86f3-492f7b9ab7b8.png">

Can you try running the [pytorch_synthetic_benchmark.py](https://github.com/horovod/horovod/blob/master/examples/pytorch_synthetic_benchmark.py) with both 1 and 4 GPUs and let me know what numbers you get?  Also, to clarify, the 4 GPUs you ran with before are on a single host, correct?

**Comment by CXMANDTXW** [comment]
@tgaddair thanks for response.
Yes, they are on a single host.
And the following is the results of p_s_b.py:
[1,0]<stdout>:Model: resnet50
[1,0]<stdout>:Batch size: 32
[1,0]<stdout>:Number of GPUs: 4
[1,0]<stdout>:Running warmup...
[1,0]<stdout>:Running benchmark...
[1,0]<stdout>:Iter #0: 73.8 img/sec per GPU
[1,0]<stdout>:Iter #1: 72.7 img/sec per GPU
[1,0]<stdout>:Iter #2: 72.2 img/sec per GPU
[1,0]<stdout>:Iter #3: 70.5 img/sec per GPU
[1,0]<stdout>:Iter #4: 73.9 img/sec per GPU
[1,0]<stdout>:Iter #5: 74.3 img/sec per GPU
[1,0]<stdout>:Iter #6: 75.8 img/sec per GPU
[1,0]<stdout>:Iter #7: 69.4 img/sec per GPU
[1,0]<stdout>:Iter #8: 72.9 img/sec per GPU
[1,0]<stdout>:Iter #9: 73.5 img/sec per GPU
[1,0]<stdout>:Img/sec per GPU: 72.9 +-3.5
[1,0]<stdout>:Total img/sec on 4 GPU(s): 291.6 +-13.8
========================================================
[1,0]<stdout>:Model: resnet50
[1,0]<stdout>:Batch size: 32
[1,0]<stdout>:Number of GPUs: 1
[1,0]<stdout>:Running warmup...
[1,0]<stdout>:Running benchmark...
[1,0]<stdout>:Iter #0: 194.5 img/sec per GPU
[1,0]<stdout>:Iter #1: 194.3 img/sec per GPU
[1,0]<stdout>:Iter #2: 193.9 img/sec per GPU
[1,0]<stdout>:Iter #3: 194.3 img/sec per GPU
[1,0]<stdout>:Iter #4: 194.1 img/sec per GPU
[1,0]<stdout>:Iter #5: 193.9 img/sec per GPU
[1,0]<stdout>:Iter #6: 193.4 img/sec per GPU
[1,0]<stdout>:Iter #7: 193.4 img/sec per GPU
[1,0]<stdout>:Iter #8: 193.3 img/sec per GPU
[1,0]<stdout>:Iter #9: 193.4 img/sec per GPU
[1,0]<stdout>:Img/sec per GPU: 193.9 +-0.9
[1,0]<stdout>:Total img/sec on 1 GPU(s): 193.9 +-0.9

**Comment by tgaddair** [comment]
Thanks for the quick response!  Looks like your scaling efficiency is about 37.5% ((291.6/193.9)/4), which is not great.  

I noticed you're not using NCCL, any particular reason?  I suspect a big part of the issue is that you're doing forward/backward passes on GPU with CUDA, but performing allreduce over CPU (host memory) with MPI.  If you were to install NCCL, I suspect you would see a significant boost.

**Comment by CXMANDTXW** [comment]
I thought that in NCCL and MPI, I just need one of them. My bad,now I will try to use NCCL, and see what the result will be.
No matter what happend, I will notify you immediately.Thank you for your great advice!

**Comment by CXMANDTXW** [comment]
Yes,NCCL Yes!
Now horovod performs very well!

**Comment by tgaddair** [comment]
Excellent!  Happy to hear that!
